tamr_unify_client/dataset/collection.py

```diff
@@ -89,4 +91,119 @@ def create(self, creation_spec):
+        :type dataset_name: str
+        :param ignore_nan: Whether to convert `NaN` values to `null` before upserting records to Unify. If `False` and
+            `NaN` is in `df`, this function will fail. Optional, default is `True`.
+        :type ignore_nan: bool
+        :returns: The newly created dataset.
+        :rtype: :class:`~tamr_unify_client.dataset.resource.Dataset`
+        :raises KeyError: If `primary_key_name` is not a column in `df`.
+        :raises CreationError: If a step in creating the dataset fails.
+        """
+        if primary_key_name not in df.columns:
+            raise KeyError(f"{primary_key_name} is not an attribute of the data")
+
+        creation_spec = {"name": dataset_name, "keyAttributeNames": [primary_key_name]}
+        try:
+            dataset = self.create(creation_spec)
+        except HTTPError as e:
+            raise CreationError(CreationError.NOT_CREATED, e)
+        # after this point, if a request fails, try to undo the change by deleting this dataset
+
+        attributes = dataset.attributes
+        for col in df.columns:
+            if col == primary_key_name:
+                # this attribute already exists, so don't create it again
+                continue
+
+            attr_spec = {
+                "name": col,
+                "type": {"baseType": "ARRAY", "innerType": {"baseType": "STRING"}},
+            }
+            try:
+                attributes.create(attr_spec)
+            except HTTPError as e:
+                self._handle_creation_failure(dataset, CreationError.ATTRIBUTE_ERROR, e)
+
+        records = df.to_dict(orient="records")
+        try:
+            response = dataset.upsert_records(
+                records, primary_key_name, ignore_nan=ignore_nan
+            )
+        except HTTPError as e:
+            self._handle_creation_failure(dataset, CreationError.RECORD_ERROR, e)
+
+        if not response["allCommandsSucceeded"]:
+            self._handle_creation_failure(
+                dataset,
+                CreationError.RECORD_VALIDATION_ERROR,
+                response["validationErrors"],
+            )
+
+        return dataset
+
+    def _handle_creation_failure(self, dataset, error_type, error):
+        """Attempts to make create_from_dataframe atomic by deleting the created dataset in the event of later failure.
+        However, this does not guarantee atomicity: if the request to delete the dataset fails, it will not retry.
+
+        :param dataset: The created dataset to delete.
+        :type dataset: :class:`~tamr_unify_client.dataset.resource.Dataset`
+        """
+        try:
+            dataset.delete()
+        except HTTPError as e:
+            raise CreationError(CreationError.DELETION_ERROR, e)
+        raise CreationError(error_type, error)
+
     # super.__repr__ is sufficient
+
+
+class CreationError(RuntimeError):
```

Comment:
Python official docs recommend that [all user-defined exceptions be derived from `Exception`](https://docs.python.org/3/library/exceptions.html#Exception):

```suggestion
class CreationError(Exception):
```